x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc;
+
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authc.service.ServiceAccountSettings;
+import org.elasticsearch.xpack.core.security.authz.store.RoleReference;
+import org.elasticsearch.xpack.core.security.user.AnonymousUser;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.xpack.core.security.authc.Authentication.VERSION_API_KEY_ROLES_AS_BYTES;
+import static org.elasticsearch.xpack.core.security.authc.AuthenticationField.API_KEY_LIMITED_ROLE_DESCRIPTORS_KEY;
+import static org.elasticsearch.xpack.core.security.authc.AuthenticationField.API_KEY_ROLE_DESCRIPTORS_KEY;
+
+/**
+ * A subject is a more generic concept similar to user and associated to the current authentication.
+ * It is more generic than user because it can also represent API keys and service accounts.
+ * It also contains authentication level information, e.g. realm and metadata so that it can answer
+ * queries in a better encapsulated way.
+ */
+public class Subject {
+
+    public enum Type {
+        USER,
+        API_KEY,
+        SERVICE_ACCOUNT,
+    }
+
+    private final Version version;
+    private final User user;
+    private final Authentication.RealmRef realm;
+    private final Type type;
+    private final Map<String, Object> metadata;
+
+    public Subject(User user, Authentication.RealmRef realm) {
+        this(user, realm, Version.CURRENT, Map.of());
+    }
+
+    public Subject(User user, Authentication.RealmRef realm, Version version, Map<String, Object> metadata) {
+        this.version = version;
+        this.user = user;
+        this.realm = realm;
+        // Realm can be null for run-as user if it does not exist. Pretend it is a user and it will be rejected later in authorization
+        // This is to be consistent with existing behaviour.
+        if (realm == null) {
+            this.type = Type.USER;
```

Comment:
This is a bit odd, but it is to retain the existing error reporting behaviour. We could argue that the existing error message is misleading anyway because it says something like
> action [...] is unauthorized for user [elastic] run as [foo] with roles [], this action is granted by the cluster privileges [...]

which feels like the fix is to grant some roles to the user `foo` but the actual problem is that user `foo` does not exist. I think we should fix it. But it should be a separate PR.